File: library/src/main/java/me/shkschneider/skeleton/java/AlphanumComparator.java
package me.shkschneider.skeleton.java;

import java.util.Comparator;

// <http://www.davekoelle.com/files/AlphanumComparator.java>
public class AlphanumComparator implements Comparator<String> {

    private boolean isDigit(char c) {
        return (c >= '0' && c <= '9');
    }

    private String getChunk(final String s, final int length, int marker) {
        final StringBuilder chunk = new StringBuilder();
        char c = s.charAt(marker);
        chunk.append(c);
        marker++;
        if (isDigit(c)) {
            while (marker < length) {
                c = s.charAt(marker);
                if (! isDigit(c)) {
                    break ;
                }
                chunk.append(c);
                marker++;
            }
        }
        else {
            while (marker < length) {
                c = s.charAt(marker);
                if (isDigit(c)) {
                    break ;
                }
                chunk.append(c);
                marker++;
            }
        }
        return chunk.toString();
    }

    @Override
    public int compare(final String s1, final String s2) {
        int m1 = 0;
        int m2 = 0;
        final int l1 = s1.length();
        final int l2 = s2.length();
        while (m1 < l1 && m2 < l2) {
            String thisChunk = getChunk(s1, l1, m1);
            m1 += thisChunk.length();
            String thatChunk = getChunk(s2, l2, m2);
            m2 += thatChunk.length();
            int result;
            if (isDigit(thisChunk.charAt(0)) && isDigit(thatChunk.charAt(0))) {
                final int thisChunkLength = thisChunk.length();
                result = thisChunkLength - thatChunk.length();
                if (result == 0) {
                    for (int i = 0; i < thisChunkLength; i++) {
                        result = thisChunk.charAt(i) - thatChunk.charAt(i);
                        if (result != 0) {
                            return result;
                        }
                    }
                }
            }
            else {
                result = thisChunk.compareTo(thatChunk);
            }
            if (result != 0) {
                return result;
            }
        }
        return (l1 - l2);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/java/AlphanumComparator.kt
package me.shkschneider.skeleton.java

// <https://gist.github.com/MichaelRocks/1b94bb44c7804e999dbf31dac86955ec>
class AlphanumComparator : Comparator<String> {

    override fun compare(s1: String, s2: String): Int {
        var thisMarker = 0
        var thatMarker = 0
        val s1Length = s1.length
        val s2Length = s2.length
        while (thisMarker < s1Length && thatMarker < s2Length) {
            val thisChunk = getChunk(s1, s1Length, thisMarker)
            thisMarker += thisChunk.length
            val thatChunk = getChunk(s2, s2Length, thatMarker)
            thatMarker += thatChunk.length
            // If both chunks contain numeric characters, sort them numerically.
            var result: Int
            if (thisChunk[0].isDigit() && thatChunk[0].isDigit()) {
                // Simple chunk comparison by length.
                val thisChunkLength = thisChunk.length
                result = thisChunkLength - thatChunk.length
                // If equal, the first different number counts.
                if (result == 0) {
                    for (i in 0 until thisChunkLength - 1) {
                        result = thisChunk[i] - thatChunk[i]
                        if (result != 0) {
                            return result
                        }
                    }
                }
            } else {
                result = thisChunk.compareTo(thatChunk)
            }
            if (result != 0) {
                return result
            }
        }
        return s1Length - s2Length
    }

    private fun getChunk(string: String, length: Int, marker: Int): String {
        var current = marker
        val chunk = StringBuilder()
        var c = string[current]
        chunk.append(c)
        current++
        if (c.isDigit()) {
            while (current < length) {
                c = string[current]
                if (!c.isDigit()) {
                    break
                }
                chunk.append(c)
                current++
            }
        } else {
            while (current < length) {
                c = string[current]
                if (c.isDigit()) {
                    break
                }
                chunk.append(c)
                current++
            }
        }
        return chunk.toString()
    }

}
